README.md
# Python3 Scone client module

A simple python3 module to interact with a scone-server.

Basic example:

    >>> from scone_client import SconeClient
    >>> client = SconeClient(host='localhost', port=6517)  # default params

    >>> client.sentence('(new-indv {Dumbo} {elephant})')
    '{Dumbo}'

    >>> client.predicate('(is-x-a-y? {Dumbo} {mammal})')
    'YES'

    >>> client.predicate('(is-x-a-y? {Dumbo} {air transport})')
    'MAYBE'

    >>> client.sentence('(new-is-a {Dumbo} {bird})')
    Traceback (most recent call last):
	  ...
      File "...", line 65, in check_error
        raise SconeError(msg)
    scone_client.SconeError: {Dumbo} cannot be a {bird}.

    >>> client.close()


The results for methods `query` and `predicate` are cached to improving performance. Any
invocation to method `sentence` invalidates de cache.


debian/changelog
scone-client (0.20161117.2-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Mon, 21 Nov 2016 10:19:49 +0000

scone-client (0.20161117.1-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Thu, 17 Nov 2016 12:19:09 +0000

scone-client (0.20161117-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Thu, 17 Nov 2016 12:09:42 +0000

scone-client (0.20161115-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Tue, 15 Nov 2016 19:33:11 +0000

scone-client (0.20161112-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Sat, 12 Nov 2016 21:48:29 +0000

scone-client (0.20161109.3-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Wed, 09 Nov 2016 22:49:42 +0000

scone-client (0.20161109.2-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Wed, 09 Nov 2016 14:09:03 +0000

scone-client (0.20161105.2-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Sat, 05 Nov 2016 14:15:18 +0000

scone-client (0.20161025.2-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Tue, 25 Oct 2016 15:35:28 +0100

scone-client (0.20161025.1-1) unstable; urgency=low

  * New release

 -- David Villa Alises <David.Villa@gmail.com>  Tue, 25 Oct 2016 15:09:55 +0100

scone-client (0.20161025-1) unstable; urgency=low

  * Initial release

 -- David Villa Alises <David.Villa@gmail.com>  Tue, 25 Oct 2016 14:44:00 +0100


test/.dir-locals.el
;;; Directory Local Variables
;;; For more information see (info "(emacs) Directory Variables")

((nil . ((nose-global-name . "nosetests3"))))


scone_client.py
#!/usr/bin/python3
# -*- coding: utf-8; mode: python -*-

import socket

BUFFERSIZE = 8192
encoding = 'utf-8'
PROMPT = b'[PROMPT]'
SCONE_ERROR = '*****SCONE-ERROR*****'

__all__ = 'SconeClient SconeError'.split()


class SconeError(Exception):
    pass


class SconeClient:
    def __init__(self, host='localhost', port=6517):
        self.sock = socket.socket()
        self.sock.connect((host, port))
        self.buffer = b''

    def close(self):
        self.sock.shutdown(socket.SHUT_RDWR)
        self.sock.close()

    def get_line(self):
        def next_line(where):
            line, self.buffer = where.split(b'\n', 1)
            return line

        if b'\n' in self.buffer:
            return next_line(self.buffer)

        data = bytes()
        while b'\n' not in data:
            data += self.sock.recv(BUFFERSIZE)

        line = next_line(data)
        # print("get-line rest:", self.buffer)
        return line

    def send(self, sentence):
        assert self.get_line() == PROMPT
        sentence = sentence.strip() + '\n'
        self.sock.sendall(sentence.encode(encoding))
        response = self.get_line().decode(encoding)
        self.check_error(response)
        return response

    def predicate(self, sentence):
        response = self.send(sentence)

        assert response in ['YES', 'NO', 'MAYBE']
        return response

    def query(self, sentence):
        return self.send(sentence)

    def check_error(self, response):
        if response.startswith(SCONE_ERROR):
            assert self.get_line() == b'NIL'
            msg = response.split('Error:')[1]
            raise SconeError(msg)


scone_client/__init__.py
#!/usr/bin/python3
# -*- coding: utf-8; mode: python -*-

import socket
from functools import lru_cache
import logging
import uuid

BUFFERSIZE = 8192
encoding = 'utf-8'
PROMPT = b'[PROMPT]'
SCONE_ERROR = '*****SCONE-ERROR*****'

__all__ = 'SconeClient SconeError'.split()


class SconeError(Exception):
    def __eq__(self, other):
        return isinstance(other, SconeError) and self.args == other.args

    def __hash__(self):
        return hash(id(self))


class SconeClient:
    def __init__(self, host='localhost', port=6517):
        self.sock = socket.socket()
        self.sock.connect((host, port))
        self.buff = b''

    def close(self):
        self.sock.shutdown(socket.SHUT_RDWR)
        self.sock.close()

    def get_line(self):
        def next_line(where):
            line, self.buff = where.split(b'\n', 1)
            return line

        if b'\n' in self.buff:
            return next_line(self.buff)

        data = bytes()
        while b'\n' not in data:
            data += self.sock.recv(BUFFERSIZE)

        line = next_line(data)
        return line

    def get_reply(self):
        reply = self.get_line()
        if not reply.startswith(b'('):
            return reply

        while not reply.endswith(b')'):
            reply += self.get_line()

        return reply

    @lru_cache(maxsize=512)
    def send(self, sentence):
        line = self.get_line()
        assert line == PROMPT, line

        sentence = sentence.strip()
        if sentence[0] != '(' or sentence[-1] != ')':
            raise SconeError("Bad input format: '{}'".format(sentence))

        logging.debug("S <- '{}'".format(sentence))

        sentence += '\n'
        self.sock.sendall(sentence.encode(encoding))
        response = self.get_reply().decode(encoding)
        logging.debug("S -> '{}'".format(response))

        self.check_error(response)
        return response

    def predicate(self, sentence):
        response = self.send(sentence)

        if response not in ['YES', 'NO', 'MAYBE']:
            reason = "Sentence '{}' was not a predicate: '{}'".format(
                sentence, response)
            raise SconeError(reason)

        return response

    def query(self, sentence):
        return self.send(sentence)

    def sentence(self, sentence):
        self.send.cache_clear()
        return self.send(sentence)

    def multi_sentence(self, sentences):
        checkpoint = 'checkpoint-{}'.format(uuid.uuid1())
        self.sentence('(new-indv {%s} {thing})' % checkpoint)

        retval = []
        for sentence in [x.strip() for x in sentences.split('\n')]:
            if not sentence:
                continue

            try:
                retval.append(self.send(sentence))
            except SconeError as e:
                self.sentence('(remove-elements-after {%s})' % checkpoint)
                self.sentence('(remove-last-element)')

                reason = "The sentence '{}' raises '{}'".format(sentence, str(e))
                raise SconeError(reason)

        return retval

    def check_error(self, response):
        if response.startswith(SCONE_ERROR):
            reply = self.get_line()
            assert reply == b'NIL', "reply was '{}'".format(reply)
            msg = response.split('Error:')[1].strip('.')
            raise SconeError(msg)
